src/idom/utils.py

```diff
@@ -63,80 +68,130 @@ def html_to_vdom(source: str, *transforms: _ModelTransform) -> Dict[str, Any]:
-        """Reset the state of the parser"""
-        self._node_stack: List[Dict[str, Any]] = []
-        super().reset()
-
-    def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> None:
-        new = self._make_vdom(tag, dict(attrs))
-        current = self._node_stack[-1]
-        current["children"].append(new)
-        self._node_stack.append(new)
-
-    def handle_endtag(self, tag: str) -> None:
-        del self._node_stack[-1]
-
-    def handle_data(self, data: str) -> None:
-        self._node_stack[-1]["children"].append(data)
-
-    @staticmethod
-    def _make_vdom(tag: str, attrs: Dict[str, Any]) -> Dict[str, Any]:
-        if "style" in attrs:
-            style = attrs["style"]
-            if isinstance(style, str):
-                style_dict = {}
-                for k, v in (part.split(":", 1) for part in style.split(";") if part):
-                    title_case_key = k.title().replace("-", "")
-                    camel_case_key = title_case_key[:1].lower() + title_case_key[1:]
-                    style_dict[camel_case_key] = v
-                attrs["style"] = style_dict
-        return {"tagName": tag, "attributes": attrs, "children": []}
+    # This will recursively call _etree_to_vdom() on all children
+    children = _generate_vdom_children(node, transforms)
+
+    # Convert the lxml node to a VDOM dict
+    attributes = dict(node.items())
+    key = attributes.pop("key", None)
+
+    if hasattr(idom.html, node.tag):
+        vdom = getattr(idom.html, node.tag)(attributes, *children, key=key)
+    else:
+        vdom: VdomDict = {"tagName": node.tag}
+        if children:
+            vdom["children"] = children
+        if attributes:
+            vdom["attributes"] = attributes
+        if key is not None:
+            vdom["key"] = key
+
+    # Perform any necessary mutations on the VDOM attributes to meet VDOM spec
+    _mutate_vdom(vdom)
+
+    # Apply any provided transforms.
+    for transform in transforms:
+        vdom = transform(vdom)
+
+    return vdom
+
+
+def _mutate_vdom(vdom: VdomDict):
+    """Performs any necessary mutations on the VDOM attributes to meet VDOM spec.
+
+    Currently, this function only transforms the ``style`` attribute into a dictionary whose keys are
+    camelCase so as to be renderable by React.
+
+    This function may be extended in the future.
+    """
+    # Determine if the style attribute needs to be converted to a dict
+    if (
+        "attributes" in vdom
+        and "style" in vdom["attributes"]
+        and isinstance(vdom["attributes"]["style"], str)
+    ):
+        # Convert style attribute from str -> dict with camelCase keys
+        vdom["attributes"]["style"] = {
+            _hypen_to_camel_case(key.strip()): value.strip()
+            for key, value in (
+                part.split(":", 1)
+                for part in vdom["attributes"]["style"].split(";")
+                if ":" in part
+            )
+        }
```

Comment:
Presently the types for `VdomDict` try to be immutable where possible. Whether we should allow them to be mutable can revisited later.

```suggestion
attributes = vdom.get("attributes")
    if (
        attributes
        and "style" in attributes
        and isinstance(attributes["style"], str)
    ):
        # Convert style attribute from str -> dict with camelCase keys
        vdom["attributes"] = {
            **attributes,
            "style": {
                hypen_to_camel_case(key.strip()): value.strip()
                for key, value in (
                    part.split(":", 1)
                    for part in vdom["attributes"]["style"].split(";")
                    if ":" in part
                )
            }
        }
```

Reply:
Recursively recreating every styled node like this really sucks, especially for DOM trees with a lot of nodes that are all styled.

Additionally, I don't think there's inherently a reason for VdomDict to be immutable?

Reply:
To clarify, `VdomAttributes` can be mutable. It's just that it doesn't have to be. Thus, MyPy is complaining here because, in the case it were not mutable, it would not be safe to assume we can assign to it. So it would be more accurate to say that `VdomAttribute` may be immutable.

Instead of the solution above, it would be better to just convince MyPy that what we're doing is in fact safe using an `assert` statement:

```suggestion
    if (
        "attributes" in vdom
        and "style" in vdom["attributes"]
        and isinstance(vdom["attributes"]["style"], str)
    ):
        # convince type checker that it's safe to mutate attributes
        assert isinstance(vdom["attributes"], dict)
        # Convert style attribute from str -> dict with camelCase keys
        vdom["attributes"]["style"] = {
            _hypen_to_camel_case(key.strip()): value.strip()
            for key, value in (
                part.split(":", 1)
                for part in vdom["attributes"]["style"].split(";")
                if ":" in part
            )
        }
```